File: DB/MainframeServices/MQR.Services/ServiceModel/QueryResult.cs
using System.Xml;
using System.Xml.Schema;
using System.Xml.Serialization;
using Microsoft.VisualBasic.CompilerServices;

namespace MQR.WebAPI.ServiceModel;

public class QueryResult : IXmlSerializable
{
   public QueryResultSection[] QueryResultSections { set; get; } = [];
      
   public DateTime MFReponseTime { get; set; } 
      
   public XmlSchema GetSchema()
   {
      return null;
   }

   public void ReadXml(XmlReader reader)
   {
      var sections = new List<QueryResultSection>();
      MFReponseTime = Convert.ToDateTime(reader.GetAttribute("MFReponseTime"));
      reader.Read();
         
      while (Operators.CompareString(reader.Name, "QueryResultSection", false) == 0 & reader.NodeType == XmlNodeType.Element)
      {
         var rows = new List<QueryResultRow>();
         var newQueryResultSection = new QueryResultSection
         {
            Identifier = reader.GetAttribute("Identifier")
         };
         reader.Read();
            
         while (Operators.CompareString(reader.Name, "QueryResultRow", false) == 0 & reader.NodeType == XmlNodeType.Element)
         {
            var fields = new List<QueryResultField>();
            var newQueryResultRow = new QueryResultRow
            {
               Identifier = reader.GetAttribute("Identifier"),
               FullLine = reader.GetAttribute("FullLine")
            };
            reader.Read();
               
            while (Operators.CompareString(reader.Name, "QueryResultField", false) == 0 & reader.NodeType == XmlNodeType.Element)
            {
               var newQueryResultField = new QueryResultField
               {
                  Identifier = reader.GetAttribute("Identifier"),
                  Value = reader.GetAttribute("Value")
               };
               reader.Read();
               
               fields.Add(newQueryResultField);
            }
            
            newQueryResultRow.Fields = fields.ToArray();
            rows.Add(newQueryResultRow);     
         }
            
         newQueryResultSection.Rows = rows.ToArray();
         sections.Add(newQueryResultSection);
      }
         
      QueryResultSections = sections.ToArray();
   }

   public void WriteXml(XmlWriter writer)
   {
      writer.WriteAttributeString("MFReponseTime", MFReponseTime.ToString("yyyy-MM-ddTHH:mm:ssz"));
         
      writer.WriteStartElement("QueryResultSections");   
      foreach (var section in QueryResultSections)
      {
         writer.WriteStartElement("QueryResultSection");
         writer.WriteAttributeString("Identifier", section.Identifier);
         foreach (var row in section.Rows)
         {
            writer.WriteStartElement("QueryResultRow");
            writer.WriteAttributeString("Identifier", row.Identifier);
            writer.WriteAttributeString("FullLine", row.FullLine);
            foreach (var field in row.Fields)
            {
               writer.WriteStartElement("QueryResultField");
               writer.WriteAttributeString("Identifier", field.Identifier);
               writer.WriteAttributeString("Value", field.Value);
               writer.WriteEndElement();
            }
            writer.WriteEndElement();
         }
         writer.WriteEndElement();
      }
      writer.WriteEndElement();   
   }
}

File: DB/MainframeServices/EWS.MQR.Services/Model/QueryResult/QueryResult.vb
Imports EWS.XML
Imports EWS.XML.XMLUtilities


<Serializable()> _
Public Class QueryResult : Inherits XMLBase

   Private m_Sections As New QueryResultSections
   Private m_MFReponseTime As Date = Date.MinValue

   Public Sub New()

   End Sub

   Public Sub New(ByVal MFReponseTime As Date)
      m_MFReponseTime = MFReponseTime
   End Sub

   Public ReadOnly Property Sections() As QueryResultSections
      Get
         Return m_Sections
      End Get
   End Property

   Public ReadOnly Property MFReponseTime() As Date
      Get
         Return m_MFReponseTime
      End Get
   End Property

   Protected Overrides Sub AddMemberDataXmlToXmlWriter(ByVal Writer As System.Xml.XmlWriter)
      Writer.WriteAttributeString("MFReponseTime", MFReponseTime.ToString("yyyy-MM-ddTHH:mm:ssz"))
      m_Sections.AddXMLToWriter(Writer)
   End Sub

   Protected Overrides Sub PopulateMemberDataFromXmlReader(ByVal Reader As System.Xml.XmlReader)

      m_MFReponseTime = CDate(Reader.GetAttribute("MFReponseTime"))
      Reader.Read()
      m_Sections.PopulateFromXMLReader(Reader)
   End Sub

   Public Overrides Function ToString() As String
      Dim Records As Integer
      For Each Section As QueryResultSection In Sections.Values
         Records += Section.Rows.Count
      Next
      Return String.Format("Response has {0} sections which contain {1} Records.", Sections.Count, Records.ToString)
   End Function

   Public Sub Merge(ByVal Addition As QueryResult)
      Merge(Me, Addition)
   End Sub

   Public Shared Sub Merge(ByVal Base As QueryResult, ByVal Addition As QueryResult)
      For Each Section As QueryResultSection In Addition.Sections.Values
         If Base.Sections.ContainsKey(Section.Key) Then
            For Each Row As QueryResultRow In Section.Rows.Values
               If DirectCast(Base.Sections(Section.Key), QueryResultSection).Rows.ContainsKey(Row.Key) Then
                  For Each Field As QueryResultField In Row.Fields.Values
                     If DirectCast(DirectCast(Base.Sections(Section.Key), QueryResultSection).Rows(Row.Key), QueryResultRow).Fields.ContainsKey(Field.Key) Then
                        DirectCast(DirectCast(DirectCast(Base.Sections(Section.Key), QueryResultSection).Rows(Row.Key), QueryResultRow).Fields(Field.Key), QueryResultField).Value = Field.Value
                     Else
                        DirectCast(DirectCast(Base.Sections(Section.Key), QueryResultSection).Rows(Row.Key), QueryResultRow).Fields.Add(Field)
                     End If
                  Next
               Else
                  DirectCast(Base.Sections(Section.Key), QueryResultSection).Rows.Add(Row)
               End If
            Next
         Else
            Base.Sections.Add(Section)
         End If
      Next
   End Sub
End Class
